### ISSUE TITLE
Request support CosyVoice2

### ISSUE THREAD

**Issue by augumn** [soniqo/speech-swift#164]
希望能支持CosyVoice2，之前基于CosyVoice2微调了很多声音模型，运转良好，调查了一下转到CosyVoice3还得重新训练，，能否支持CosyVoice2

**Comment by ivan-digital** [comment]
感谢您的请求！保留已有的微调语音模型很重要——重新训练成本很高。

为了评估支持范围，想了解一些细节：

1. **使用场景**：您用 CosyVoice2 做什么？目标平台是 Mac、iOS 还是服务端？
2. **模型格式**：您的微调模型是 PyTorch（`.pt`/`.bin`）还是 ONNX？基于哪个基础 checkpoint 微调的？
3. **架构版本**：具体是哪个 CosyVoice2 变体？（例如 `CosyVoice2-0.5B`，具体仓库/checkpoint 名称）
4. **训练数据**：微调用了什么数据？（时长、语言、单说话人还是多说话人、是否特定领域？）
5. **推理后端**：需要 MLX（Mac GPU）、CoreML（Neural Engine/iOS）还是两者都需要？
6. **量化**：您的模型使用什么精度？INT4/INT8 量化推理是否可以接受？
7. **声音克隆**：您的微调模型是基于说话人嵌入的，还是完全微调的 checkpoint？

我们库中的 CosyVoice3 已经支持 DiT flow matching + HiFi-GAN 架构。如果 v2 共享足够多的架构，可能通过配置切换来同时支持两个版本，而不需要全新模块。

**Comment by augumn** [comment]
背景描述：
主要是使用CosyVoice2进行声音推理，这个项目使我燃起了在mac上推理的希望，在mac上启动一个api服务，给到业务层调用，之前我是在autodl上进行CosyVoice2的微调（flow.pt和llm.pt），然后把模型下载到我本地（windows上+小显卡）。

举例：
我要训练“张三”的声音，我会准备好素材和标记，用CosyVoice2自带的训练脚本训练出flow.pt和llm.pt,
然后命名为：
1.zhangsan-cosyvoice-llm.pt
1.zhangsan-cosyvoice-flow.pt

“zhangsan”为model_id


代码片段：
```
def inference_zero_shot(self, tts_text, prompt_text, prompt_speech_16k, zero_shot_spk_id='', stream=False, speed=1.0, text_frontend=True, model_id=''):

        if model_id and model_id != self.model_id:
            logging.info('Switching model {} to {}'.format(self.model_id, model_id))
            self.model_id = model_id
            llm_model_path, flow_model_path, hift_model_path = verify_and_get_model_paths(model_id, self.model_dir, self.multiple_self_trained_dirs)
            self.model.reload(llm_model_path, flow_model_path)
```

```
def verify_and_get_model_paths(model_id, default_pretrained_models_path, multiple_self_trained_dirs):
    hift_model_path = os.path.join(default_pretrained_models_path, 'hift.pt')

    # 初始化为 None，便于后续逻辑判断
    llm_model_path = None
    flow_model_path = None

    # 查找逻辑
    if model_id.startswith('0.default'):
        # 默认模型路径
        temp_llm = os.path.join(default_pretrained_models_path, 'llm.pt')
        temp_flow = os.path.join(default_pretrained_models_path, 'flow.pt')

        # 这里不需要立即检查 exists，留给最后统一检查，或者立即检查并报错
        llm_model_path = temp_llm
        flow_model_path = temp_flow

    else:
        # 遍历自定义目录查找
        for tmp_trained_dir in multiple_self_trained_dirs:
            # 构造路径
            temp_llm = os.path.join(tmp_trained_dir, f'{model_id}-cosyvoice-llm.pt')
            temp_flow = os.path.join(tmp_trained_dir, f'{model_id}-cosyvoice-flow.pt')

            # 如果两个文件都存在，则认定找到了，并立即跳出循环
            if os.path.exists(temp_llm) and os.path.exists(temp_flow):
                llm_model_path = temp_llm
                flow_model_path = temp_flow
                break

    # 单独检查 HiFT
    if not os.path.exists(hift_model_path):
        raise RuntimeError(f"基础模型文件缺失: {hift_model_path}")

    # 检查 LLM 和 Flow 是否找到且存在
    # 注意：如果 model_id 是 default，路径已赋值但文件可能不存在；
    # 如果是自定义 ID 且循环没找到，路径可能是 None。
    if not llm_model_path or not os.path.exists(llm_model_path):
        raise RuntimeError(f"LLM 模型未找到: model_id={model_id}, path={llm_model_path or '未在搜索路径中找到'}")

    if not flow_model_path or not os.path.exists(flow_model_path):
        raise RuntimeError(f"Flow 模型未找到: model_id={model_id}, path={flow_model_path or '未在搜索路径中找到'}")

    return llm_model_path, flow_model_path, hift_model_path
```

```
def reload(self, llm_model=None, flow_model=None, hift_model=None):
        """
        动态切换模型。
        """
        import gc  # 引入垃圾回收

        # 1. 切换前先做一次大扫除，清理之前的推理碎片
        torch.cuda.empty_cache()

        with self.lock:
            # --- 切换 LLM ---
            if llm_model is not None:
                state_dict = torch.load(llm_model, map_location='cpu')
                target_llm = getattr(self.llm, '_orig_mod', self.llm)
                target_llm.load_state_dict(state_dict, strict=True)

                self.llm.eval()  
                self.current_llm_model = llm_model

                del state_dict
                gc.collect()

            # --- 切换 Flow ---
            if flow_model is not None:

                state_dict = torch.load(flow_model, map_location='cpu')
                
                target_flow = getattr(self.flow, '_orig_mod', self.flow)
                target_flow.load_state_dict(state_dict, strict=True)
                
                self.flow.eval()
                self.current_flow_model = flow_model
                del state_dict
                gc.collect()


        # 2. 切换后再清理一次显存，确保万无一失
        torch.cuda.empty_cache()
```



1.使用场景：您用 CosyVoice2 做什么？目标平台是 Mac、iOS 还是服务端？
场景CosyVoice2+自训练模型制作语音作品，目标是mac，用swift构建一个webapi

2.模型格式：您的微调模型是 PyTorch（.pt/.bin）还是 ONNX？基于哪个基础 checkpoint 微调的？
.pt

3.架构版本：具体是哪个 CosyVoice2 变体？（例如 CosyVoice2-0.5B，具体仓库/checkpoint 名称）
CosyVoice2-0.5B

4.训练数据：微调用了什么数据？（时长、语言、单说话人还是多说话人、是否特定领域？）
30分钟左右的人声切片+文字，然后微调llm.pt和flow.pt

5.推理后端：需要 MLX（Mac GPU）、CoreML（Neural Engine/iOS）还是两者都需要？
Mac GPU,主要是用mac作为服务提供api

6.量化：您的模型使用什么精度？INT4/INT8 量化推理是否可以接受？
不太懂，问了AI，帮我修改了成bfloat16，在小显卡上提速还挺明显的（40系显卡必须开 bf16，省显存）

7.声音克隆：您的微调模型是基于说话人嵌入的，还是完全微调的 checkpoint？
不懂

**Comment by augumn** [comment]
我不太懂这些模型内部的架构推理，只是从工具层面把CosyVoice2变成了一个顺手的工具，可能很多地方是蒙的，但是到目前为止，我的这些微调出来的模型能够满足业务需要。

看到这个项目以及了解到现在apple在ai上的统一内存架构MLX，让我燃起了想试试的念头，或许会有很不错的收获。

<img width="284" height="791" alt="Image" src="https://github.com/user-attachments/assets/bf70fa58-048f-4858-aba2-dc6f9cd7dd4a" />

**Comment by augumn** [comment]
以下是让AI充分阅读我的代码以及转向MLX的技术分析，方便判断 CosyVoice2 是否值得支持：

1. 使用场景：主要是服务端 TTS / 声音克隆，不是 iOS 优先。目标是 macOS Apple Silicon，用 Swift + MLX 在 Mac 上跑 Web API 给业务调用。
2. 模型格式：当前使用的是 PyTorch `.pt` checkpoint，不是 ONNX。基于 `CosyVoice2-0.5B` 路线，实际部署会替换微调后的 `llm.pt` 和 `flow.pt`，`hift.pt` 目前复用基础模型。
3. 架构版本：我这边对应的是 CosyVoice2-0.5B，不是 CosyVoice3。
4. 训练数据：主要是单说话人，约 30 分钟左右切片 + 对应文本，中文为主。对我来说重点是保留已有可用声音资产，而不是迁移到 CV3 后重新训练。
5. 推理后端：当前最需要的是 MLX / macOS Apple Silicon。CoreML / iOS 不是第一优先级。
6. 量化/精度：目前更关心先跑通 float / fp16 / bf16 兼容路线；bf16 在小显存环境里对速度和显存都有帮助。INT4/INT8 可以后续再评估，但不是第一诉求。
7. 声音克隆方式：我的推理不是只有 speaker embedding，也不是简单 text-to-speech。实际是 CosyVoice2 的 zero-shot / prompt-conditioned 推理，运行时会同时使用 prompt text、prompt speech token、prompt speech feature、speaker embedding；同时模型本身也替换了微调后的 `llm.pt` 和 `flow.pt`。

补充一个实现上很关键的点：如果目标是真正兼容现有 CosyVoice2 微调模型，我理解这大概率不是“在当前 CosyVoice3 支持上切一个 config”就够了。因为我现在跑通的入口是 `inference_zero_shot(...)`，会读取每个 `model_id` 对应的 `prompt.wav + prompt.txt`，前端还会生成：
- `llm_prompt_speech_token`
- `flow_prompt_speech_token`
- `prompt_speech_feat`
- `llm_embedding`
- `flow_embedding`

从配置看，这条线是 `Qwen2LM + CausalMaskedDiffWithXvec / UpsampleConformerEncoder + HiFTGenerator`，24k 采样率，`speech_token_size=6561`，`token_mel_ratio=2`。所以如果要兼容现有 CV2 微调资产，可能需要一条明确的 CosyVoice2 runtime path，至少要保留 zero-shot prompt conditioning。

**Comment by ivan-digital** [comment]
Thanks for the detailed breakdown — very helpful.

Looking at the architecture differences, CosyVoice2 and CosyVoice3 share the FSQ tokenizer (6561 vocab, 24kHz) but diverge significantly on the flow model (CausalMaskedDiffWithXvec vs DiT) and vocoder (HiFT vs HiFi-GAN). Supporting v2 would require a separate runtime path, not just a config switch.

Our recommendation would be to retrain your voice models on CosyVoice3 — the training pipeline is similar (30min audio + text), and v3 has better synthesis quality. Your existing audio data and labels transfer directly.

If you'd like to add CosyVoice2 support yourself, the existing CosyVoice3 implementation in `Sources/CosyVoiceTTS/` and the conversion script `scripts/convert_cosyvoice3.py` would be a good starting point. The main work is the flow model (CausalMaskedDiff vs DiT) and vocoder (HiFT vs HiFi-GAN) — the LLM and tokenizer should mostly carry over. PRs welcome!

For now we're prioritizing the v3 pipeline and iOS integration.

---

感谢您详细的说明，非常有帮助。

从架构差异来看，CosyVoice2 和 CosyVoice3 共享 FSQ 分词器（6561词表，24kHz），但在 flow 模型（CausalMaskedDiffWithXvec vs DiT）和声码器（HiFT vs HiFi-GAN）上有较大差异。支持 v2 需要一条独立的运行时路径，而不是简单的配置切换。

我们的建议是在 CosyVoice3 上重新训练您的声音模型——训练流程类似（30分钟音频+文本），v3 的合成质量更好。您现有的音频数据和标注可以直接复用。

如果您想自己添加 CosyVoice2 支持，可以参考现有的 CosyVoice3 实现 `Sources/CosyVoiceTTS/` 和转换脚本 `scripts/convert_cosyvoice3.py`。主要工作是 flow 模型（CausalMaskedDiff vs DiT）和声码器（HiFT vs HiFi-GAN）——LLM 和分词器部分基本可以复用。欢迎提交 PR！

目前我们的重点是 v3 流程和 iOS 集成。
